File: core/src/main/java/org/apache/spark/raydp/AppMasterJavaBridge.java
package org.apache.spark.raydp;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.ray.api.Ray;
import org.apache.spark.deploy.raydp.RayAppMaster;

import java.util.*;

public class AppMasterJavaBridge {
    private RayAppMaster instance = null;

    public AppMasterJavaBridge(String properties) {
        ObjectMapper mapper = new ObjectMapper();
        TypeReference<HashMap<String, String>> typeRef
                = new TypeReference<HashMap<String, String>>() {};
        HashMap<String, String> propertiesMap = null;
        try {
            propertiesMap = mapper.readValue(properties, typeRef);
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
        for (Map.Entry<String, String> pair: propertiesMap.entrySet()) {
            System.setProperty(pair.getKey(), pair.getValue());
        }
    }

    public void createAppMaster(String extra_cp) {
        if (instance == null) {
            // init ray, we should set the config by java properties
            Ray.init();
            System.out.println(Ray.getRuntimeContext().getCurrentJobId());
            instance = new RayAppMaster(extra_cp);
        }
    }

    public String getMasterUrl() {
        if (instance == null) {
            throw new RuntimeException("You should create the RayAppMaster instance first");
        }

        return instance.getMasterUrl();
    }

    public void stop() {
        if (instance != null) {
            instance.stop();
            instance = null;
        }
    }
}

File: core/src/main/scala/org/apache/spark/deploy/raydp/AppMasterJavaBridge.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.deploy.raydp

import io.ray.api.Ray
import io.ray.runtime.config.RayConfig
import org.json4s._
import org.json4s.jackson.JsonMethods._

class AppMasterJavaBridge {
  private var instance: RayAppMaster = null

  def setProperties(properties: String): Unit = {
    implicit val formats = org.json4s.DefaultFormats
    val parsed = parse(properties).extract[Map[String, String]]
    parsed.foreach{ case (key, value) =>
      System.setProperty(key, value)
    }
    // Use the same session dir as the python side
    RayConfig.getInstance().setSessionDir(System.getProperty("ray.session-dir"))
  }

  def startUpAppMaster(extra_cp: String): Unit = {
    if (instance == null) {
      // init ray, we should set the config by java properties
      Ray.init()
      instance = new RayAppMaster(extra_cp)
    }
  }

  def getMasterUrl(): String = {
    if (instance == null) {
      throw new RuntimeException("You should create the RayAppMaster instance first")
    }
    instance.getMasterUrl()
  }

  def stop(): Unit = {
    if (instance != null) {
      instance.stop()
      instance = null
    }
  }
}
